src/kerberos/mod.rs

```diff
@@ -653,40 +654,225 @@ impl<'a> Kerberos {
+            "HTTP" => self.initialize_security_context_impl_plain(yield_point, builder).await,
+            _ => {
+                self.initialize_security_context_impl_p2p(yield_point, builder, &username, service_name)
+                    .await
+            }
+        }
+    }
+
+    pub(crate) async fn initialize_security_context_impl_plain(
+        &'a mut self,
+        yield_point: &mut YieldPointLocal,
+        builder: &'a mut crate::builders::FilledInitializeSecurityContext<'_, <Self as SspiImpl>::CredentialsHandle>,
+    ) -> Result<crate::InitializeSecurityContextResult> {
+        let credentials = builder
+            .credentials_handle
+            .as_ref()
+            .unwrap()
+            .as_ref()
+            .ok_or_else(|| Error::new(ErrorKind::WrongCredentialHandle, "No credentials provided"))?;
+
+        let (username, password, realm, cname_type) = match credentials {
+            CredentialsBuffers::AuthIdentity(auth_identity) => {
+                let username = utf16_bytes_to_utf8_string(&auth_identity.user);
+                let domain = utf16_bytes_to_utf8_string(&auth_identity.domain);
+                let password = utf16_bytes_to_utf8_string(auth_identity.password.as_ref());
+
+                let realm = get_client_principal_realm(&username, &domain);
+                let cname_type = get_client_principal_name_type(&username, &domain);
+
+                (username, password, realm, cname_type)
+            }
+        };
+        self.realm = Some(realm.clone());
+
+        let options = GenerateAsReqOptions {
+            realm: &realm,
+            username: &username,
+            cname_type,
+            snames: &[TGT_SERVICE_NAME, &realm],
+            // 4 = size of u32
+            nonce: &OsRng.gen::<[u8; 4]>(),
+            hostname: &unwrap_hostname(self.config.hostname.as_deref())?,
+            context_requirements: builder.context_requirements,
+        };
+        let kdc_req_body = generate_as_req_kdc_body(&options)?;
+
+        let pa_data_options = match credentials {
+            CredentialsBuffers::AuthIdentity(auth_identity) => {
+                let domain = utf16_bytes_to_utf8_string(&auth_identity.domain);
+                let salt = format!("{}{}", domain, username);
+
+                AsReqPaDataOptions::AuthIdentity(GenerateAsPaDataOptions {
+                    password: &password,
+                    salt: salt.as_bytes().to_vec(),
+                    enc_params: self.encryption_params.clone(),
+                    with_pre_auth: false,
+                })
+            }
+        };
+
+        let as_rep = self.as_exchange(yield_point, &kdc_req_body, pa_data_options).await?;
+
+        info!("AS exchange finished successfully.");
+
+        self.realm = Some(as_rep.0.crealm.0.to_string());
+
+        let (encryption_type, salt) = extract_encryption_params_from_as_rep(&as_rep)?;
+
+        let encryption_type = CipherSuite::try_from(encryption_type as usize)?;
```

Comment:
Use `usize::from` instead of `as` so we’re sure the conversion is lossless (`From` trait is for lossless conversions only). In practice this prevents conversions from turning into silent lossy conversions if the types of the input expressions ever change, and it’s also easier for people reading the code to know that the conversion is lossless.

```suggestion
        let encryption_type = CipherSuite::try_from(usize::from(encryption_type))?;
```